File: kinotic-grind/src/main/java/org/kinotic/grind/api/model/events/TasksDiscoveredEvent.java
package org.kinotic.grind.api.model.events;

import org.kinotic.grind.api.model.TaskRecord;

import java.util.List;

/**
 * A task produced dynamically discovered tasks at runtime, carrying the discovered subtree as
 * PENDING {@link TaskRecord}s in discovery order.
 *
 * @param taskPath the position of the task that produced the discovery
 * @param tasks    the discovered tasks, recorded PENDING
 */
public record TasksDiscoveredEvent(String taskPath, List<TaskRecord> tasks) implements JobRunEvent {
}

File: kinotic-js/workspace/packages/management-api/src/api/model/grind/events/TasksDiscoveredEvent.ts
import { JobRunEventType } from '@/api/model/grind/events/JobRunEventType'
import type { TaskRecord } from '@/api/model/grind/TaskRecord'

/**
 * The tasks of the run became known: the definition's own tasks at the start of the run, and
 * the subtree a task produced when it generates tasks at runtime.
 */
export class TasksDiscoveredEvent {

    public readonly type: JobRunEventType.TASKS_DISCOVERED = JobRunEventType.TASKS_DISCOVERED

    /**
     * The position of the task the discovered tasks belong under.
     */
    public taskPath: string = ''

    /**
     * The discovered tasks, recorded PENDING, in discovery order.
     */
    public tasks: TaskRecord[] = []

    /**
     * True when the task at taskPath produced the discovery at runtime, false for the
     * definition's own tasks at the start of the run.
     */
    public dynamic: boolean = false

}
